design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,123 @@
+
+The introduction of lambda functions will ultimately empower ksqDB users to solve more problems with less effort. Many of the UDFs we’ve seen users write are designed to apply relatively simple functionality to the elements of `Arrays` and `Maps`. Lambda functions are well-suited to solve this class of problems in a very easy and intuitive way.
+
+Lambda functions will also enable some users to use ksqlDB in environments that do not allow for the deployment of UDFs.
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
+
+```
+arg => expr
+```
+
+This expression represents the following:
+
+* `arg` is the argument passed into the scope of each invocation of the lambda function.
+* `expr` is the "body" of the lambda function. Each invocation of the lambda function will evaluate this expression and return the result.
+
+Multiple arguments should also be allowed for lambda functions. When multiple arguments are used, they should be wrapped in parentheses:
+
+```
+(x, y) => expr
+```
+
+#### Allowed expressions
+
+The body of each lambda function should simply be **any expression that could be used in a `SELECT` expression list**, with some exceptions:
+
+* Aggregate function calls (e.g. `sum`, `count`, `avg`) should not be allowed (please see *[Open questions](#open-questions)* section). Note that simple aggregations are still possible with the `reduce_*` invocation functions.
+* `SELECT` subqueries should not be allowed.
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_keys(map, k => k)` - Applies the given lambda function to each key in the input `Map`, returning a new `Map` containing the transformed keys.
+
+* `transform_values(map, v => v)` - Applies the given lambda function to each value in the input `Map`, returning a new `Map` containing the transformed values.
+
```

Comment:
What about a `transform_map` method that allows the user to convert both keys and values together, e.g. 

`transform_map(map, (k, v) => newKey, (k, v) => newValue)`

So it takes two lambda, but each lambda only returns a single value (to keep things simple).  Definitely worth passing the both key and value to each lambda.

Reply:
This seems like a good idea. The original proposal included a `transform_map` invocation function that looked like this:

```
transform_map(m, (k, v) => (x, y))
```

But adding support for returning multiple values would probably be a lot of work so I broke it into `transform_keys` and `transform_values`. I think it would make sense to support a variant that simply took two lambdas though. I'll add this to the proposal.

Reply:
@big-andy-coates do you think it would make sense to give each lambda function access to both the key and the value? e.g.,

```
transform_map(m, (k, v) => k, (k, v) => v)
```

Reply:
Would setting a new key not potentially be "undefined", ie, if two entries map to the same new key?

Reply:
> Would setting a new key not potentially be "undefined", ie, if two entries map to the same new key?

I think you're right.

If we define the order of iteration of the entries in a `MAP`, then we could support the processing of a later entry _overwriting_ a previous output entry, or support ignoring any subsequent results that clash with an existing key.

If we wish the SQL `MAP` type to be an unordered collection, choosing to keep either the first or last entry of a collisions would introduce nondeterminism. So we either say the result is _undefined_ or we throw an exception, i.e. log an error and populate the result with `NULL`.  

IMHO, we should keep the SQL `MAP` type an unordered collection, and of the two ways to handle key collisions when transforming an unordered `MAP`, the _undefined_ result seems the lesser of two evils.